### PR TITLE
[Rendering] Renderer optimization, debug utilities

### PR DESCRIPTION
**Pull request by MomoDeve**
this PR:
- eliminates unnecessary mipmap generation for render targets
- introduces separate render target what can be downsamples by generating mips
- adds debug group macro & provides descriptions for render passes (can be seen in GPU debugger, see pic below)
- modifies SSGI & SSAO to not use mipmaps during computing & during blurring (only there is one downsample after compute)
- depth pyramid is computed after draw objects with mipmaps

image

### CODE REVIEW

File: src/Core/Rendering/RenderController.cpp [hunk 1]
```diff
@@ -318,30 +330,39 @@ namespace MxEngine
...
         applyShader->SetUniform("intensity", camera.SSAO->GetIntensity());
 
         this->RenderToTexture(output, applyShader);
         std::swap(input, output);
     }
 
+    void RenderController::GenerateDepthPyramid(TextureHandle& depth)
+    {
+        MAKE_RENDER_PASS_SCOPE("RenderController::GenerateDepthPyramid()");
+        // generate depth texture mipmaps for post-processing algorithms. Replace later with hierarhical depth map
```
Comment (fall2019) [inline src/Core/Rendering/RenderController.cpp:358]: 👌

**Review by fall2019 — APPROVED** [review]

**Comment by fall2019** [comment]
image
image
Oh there was something wrong with this pr.

**Comment by MomoDeve** [comment]
@fall2019 yeah, I changed ssgi/ssao algorithms a bit. Do you think they became worse?

**Comment by fall2019** [comment]
Maybe the brighter one looks more realistic since the intensity of the directional light is very high. And also it lights up the floor with high energy but other places seems too dark.

**Pull request merged**